Lib/subprocess.py

```diff
@@ -968,12 +972,40 @@ def _execute_child(self, args, executable, preexec_fn, close_fds,
             # Process startup details
             if startupinfo is None:
                 startupinfo = STARTUPINFO()
-            if -1 not in (p2cread, c2pwrite, errwrite):
+
+            use_std_handles = -1 not in (p2cread, c2pwrite, errwrite)
+            if use_std_handles:
                 startupinfo.dwFlags |= _winapi.STARTF_USESTDHANDLES
                 startupinfo.hStdInput = p2cread
                 startupinfo.hStdOutput = c2pwrite
                 startupinfo.hStdError = errwrite
 
+            attribute_list = startupinfo.lpAttributeList
+            have_handle_list = (attribute_list and
+                                "handle_list" in attribute_list and
+                                attribute_list["handle_list"])
+
+            # If we were given an handle_list or need to create one
+            if have_handle_list or (use_std_handles and close_fds):
+                if attribute_list is None:
+                    attribute_list = startupinfo.lpAttributeList = {}
+                handle_list = attribute_list.setdefault("handle_list", [])
+
+                if use_std_handles:
+                    handle_list += [int(p2cread), int(c2pwrite), int(errwrite)]
+
+                handle_list[:] = self._filter_handle_list(handle_list)
+
+                if handle_list:
+                    if have_handle_list and not close_fds:
```

Comment:
Whoops the `if` there was wrong in a different way... `not close_fds` should have been `close_fds` and have_handle_list was pointing to have reference to handle_list which was later modified also changing have_handle_list.

The following should not generate a warning:
```python
>>> import subprocess
>>> subprocess.Popen(['notepad.exe'], stdout=subprocess.PIPE)
<subprocess.Popen object at 0x03A0FFC0>
>>> # No warning
```

The following should:
```python
>>> import os, subprocess, msvcrt
>>> f=open('bla.txt', 'w')
>>> handle = msvcrt.get_osfhandle(f.fileno())
>>> startupinfo = subprocess.STARTUPINFO()
>>> startupinfo.lpAttributeList["handle_list"] = [handle]
>>> os.set_inheritable(f.fileno(), True)
>>> subprocess.Popen(['notepad.exe'], startupinfo=startupinfo)
C:\Users\Segev\prj\python\cpython\lib\subprocess.py:1002: RuntimeWarning: startupinfo.lpAttributeList['handle_list'] overriding close_fds
  "overriding close_fds", RuntimeWarning)
<subprocess.Popen object at 0x03B0E9A0>
>>> 
```

If I remove `have_handle_list` from the if than you get a warning for the first example:
```python
>>> import subprocess
>>> subprocess.Popen(['notepad.exe'], stdout=subprocess.PIPE)
C:\Users\Segev\prj\python\cpython\lib\subprocess.py:1002: RuntimeWarning: startupinfo.lpAttributeList['handle_list'] overriding close_fds
  "overriding close_fds", RuntimeWarning)
<subprocess.Popen object at 0x02C0FFC0>
>>>
```